doc: Add trusted storage library (#13581)
nrfconnect/sdk-nrf: doc/nrf/libraries/others/trusted_storage.rst

.. trusted_storage_readme:

Trusted storage library
#######################

.. contents::
   :local:
   :depth: 2

Overview
********

The trusted storage library allows its users to securely store data.
It is based on PSA Certified Secure Storage API and provides data protection through integrity and confidentiality using AEAD algorithms.

Interfaces
==========

This trusted storage library provides two storage interfaces for use with device-protected storage:

* Internal trusted storage, designed to store critcal data that must be placed inside internal flash memory.
  Examples of assets that require this kind of storage are replay protection values for external storage and keys for use by components of the PSA Root of Trust.
* Protected storage, designed to store all other critical data that do not need to be stored inside internal flash memory.

Backends
========

The following backends are available in the trusted storage library:

> SECURE_STORAGE_BACKEND_AEAD

Author: Are we changing all instances of SECURE_STORAGE to TRUSTED_STORAGE now?

Reviewer: I renamed to to encrypted storage, in #13833

   The ``SECURE_STORAGE_BACKEND_AEAD`` backend handles encryption, authentication or other means of validation of the stored data.

>    It is responsible for what is happening with the data before and after they are read from or written to non-volatile storage (flash memory).

Author: changed to non-volatile memory :)

   The trusted storage library implements only ``SECURE_STORAGE_BACKEND_AEAD`` as a secure implementation backend, but it has support for adding other secure implementation backends to provide various levels of trust, depending on the device security features.

SECURE_STORAGE_STORAGE_BACKEND_SETTINGS
   This backend handles how the data are written to non-volatile storage.
   It uses :ref:`Zephyr's settings subsystem <zephyr:settings_api>` for storing of data.

   The trusted storage library implements only ``SECURE_STORAGE_STORAGE_BACKEND_SETTINGS`` as a storage backend, but it has support for adding other memory types for storage.

Requirements
************

> 

Author: Need more details in this section

Zephyr's settings subsystem has to be enabled for use?

Configuration
*************

Set the Kconfig option :kconfig:option:`CONFIG_SECURE_STORAGE` to enable the trusted storage library.

Once the library is enabled, two configuration options have to be set, :kconfig:option:`CONFIG_SECURE_STORAGE_BACKEND` and :kconfig:option:`CONFIG_SECURE_STORAGE_STORAGE_BACKEND`.

Use the Kconfig option :kconfig:option:`CONFIG_SECURE_STORAGE_BACKEND` to define the backend that handles encryption and authentication.
If this Kconfig option is set, the configuration defaults to the only currently available option :kconfig:option:`CONFIG_SECURE_STORAGE_BACKEND_AEAD` to use an AEAD scheme for encryption and authentication of stored data.

Use the Kconfig option :kconfig:option:`CONFIG_SECURE_STORAGE_STORAGE_BACKEND` to define the backend that handles how the data are written to and from the non-volatile storage.
If this Kconfig option is set, the configuration defaults to the only currently available option :kconfig:option:`CONFIG_SECURE_STORAGE_STORAGE_BACKEND_SETTINGS` to use Zephyr's settings subsystem.

The following options are used to configure the AEAD backend and its behavior:

> * :kconfig:option:`CONFIG_SECURE_STORAGE_BACKEND_AEAD_MAX_DATA_SIZE` - Defines the maximum data storage size for the AEAD backend (256 as default value).

Author: I see that some Kconfig options have BACKEND_AEAD and some AEAD_BACKEND in their name. Is this intentional? And if yes, why? :)

Reviewer: also renamed it in #13833

  * :kconfig:option:`CONFIG_SECURE_STORAGE_BACKEND_AEAD_NONCE_PSA_SEED_COUNTER` - Selects the PSA Crypto for nonce initial random seed and a counter incrementing nonce for each AEAD encryption.
    This is the default option for the AEAD nonce.
  * :kconfig:option:`CONFIG_SECURE_STORAGE_BACKEND_AEAD_NONCE_CUSTOM` - Selects a custom implementation for AEAD nonce provider.
* :kconfig:option:`CONFIG_SECURE_STORAGE_BACKEND_AEAD_KEY` - Selects what implementation provides the AEAD keys.
  You can choose one of the following values when this configuration option is set:

  * :kconfig:option:`CONFIG_SECURE_STORAGE_BACKEND_AEAD_KEY_DERIVE_FROM_HUK` - Selects HUK to derive a key based on the UID file.
    This is the default selection for the AEAD key provider, and also the only really secure option.
  * :kconfig:option:`CONFIG_SECURE_STORAGE_BACKEND_AEAD_KEY_HASH_UID` - Selects the use of SHA-256 of the UID file as the key.
    This option is not as secure as when using HUKs for key derivation.
	Use this option only when HUK is not possible to use.
  * :kconfig:option:`CONFIG_SECURE_STORAGE_AEAD_BACKEND_KEY_CUSTOM` - Selects a custom implementation for the AEAD key provider.

Usage
*****

> 

Author: Need input on usage of the library.

> Samples using the library?

Author: Are we planning on adding a sample that showcases the library?

> Limitations?

Author: I guess this is not a library limitation or a shortcoming.

Dependencies?

Do these have to be mentioned:

Error codes - Header file: :file:`subsys/secure_storage/include/psa/error.h`

> General definitions - Header file: :file:`subsys/secure_storage/include/psa/storage_common.h`

Author: Some open questions that I left in the file for the moment.

.. trusted_storage_readme:

Trusted storage
###############

.. contents::
   :local:
   :depth: 2

Overview
********

> The trusted storage library enables its users to provide integrity, confidentiality and authenticity of stored data using AEAD algorithms or cryptographic hash.

Reviewer: Maybe we add here:
Without using our Platform Root of Trust (PRoT) / TF-M.

The library implements the PSA Certified Secure Storage API.

Architecture

> ============

Reviewer: Sorry for the miss understanding was more thinking about the Architecture of this subsystem.

Feel free to pre-hrase this nicely.

The subsystem in designed in a modular way to allow a degree of customization. 
This is realized by using backends for the different tasks within the subsystem. 
There is the choice of the `ENCRYPTED_STORAGE_BACKEND` that responsible to modify the assets before they get forwarded to the `ENCRYPTED_STORAGE_STORAGE_BACKEND` which takes care of storing the data in Non-volatile storage. 

The provided backends are also modular to customize which crypto implementation and the key, nonce are provided.

Details about the single backends and their choices are described in -> Backends 

The figure XX gives and overview of the default architecture of the subsystem.

* ``SECURE_STORAGE_BACKEND`` that defines a backend for handling of encryption, authentication or other means of validation of the stored data.
  It is responsible for what is happening with the data before and after they are read from or written to non-volatile storage.
* ``SECURE_STORAGE_STORAGE_BACKEND`` that defines a backend that handles how the data are written to non-volatile storage.

The trusted storage library provides ``SECURE_STORAGE_BACKEND_AEAD`` and ``SECURE_STORAGE_STORAGE_BACKEND_SETTINGS`` as default backends for the ``SECURE_STORAGE_BACKEND`` and the ``SECURE_STORAGE_STORAGE_BACKEND`` respectively.

Backends
========

The following backends are used in the trusted storage library:

SECURE_STORAGE_BACKEND_AEAD
   Uses an AEAD scheme to provide integrity, confidentiality and authenticity.
   The trusted storage library provides the ``SECURE_STORAGE_BACKEND_AEAD`` backend, but it has support for adding other secure implementation backends to provide various levels of trust, depending on the device security features.

   Uses a separate backend to perform the AEAD operation, set by the Kconfig option :kconfig:option:`SECURE_STORAGE_AEAD_BACKEND_CRYPTO`.

>    The nonce and the key for the AEAD operation are provided by separate backends, allowing a custom application.

Reviewer: ```suggestion
   The nonce and the key for the AEAD operation are provided by separate backends, allowing customization 
```
Else it sounds a bit wired what custom application is this allowing? 
Alternatively `allowing a custom implementation` could be also an option.

:kconfig:option:`CONFIG_SECURE_STORAGE_BACKEND_AEAD_NONCE`
   Selects what implementation provides AEAD nonce.
   You can choose one of the following values when this configuration option is set:

   * :kconfig:option:`CONFIG_SECURE_STORAGE_BACKEND_AEAD_NONCE_PSA_SEED_COUNTER` - Selects the PSA Crypto for nonce initial random seed and a counter incrementing nonce for each AEAD encryption.
     This is the default option for the AEAD nonce.
   * :kconfig:option:`CONFIG_SECURE_STORAGE_BACKEND_AEAD_NONCE_CUSTOM` - Selects a custom implementation for AEAD nonce provider.

:kconfig:option:`CONFIG_SECURE_STORAGE_BACKEND_AEAD_KEY`
   Selects what implementation provides the AEAD keys.
   You can choose one of the following values when this configuration option is set:

   * :kconfig:option:`CONFIG_SECURE_STORAGE_BACKEND_AEAD_KEY_DERIVE_FROM_HUK` - Selects HUK to derive a key based on the UID file.
     This is the default selection for the AEAD key provider, and also the only really secure option.
   * :kconfig:option:`CONFIG_SECURE_STORAGE_BACKEND_AEAD_KEY_HASH_UID` - Selects the use of SHA-256 of the UID file as the key.

>      This option is not as secure as when using HUKs for key derivation.

Reviewer: To make it clear we should add that this will only provide integrity

     Use this option only when HUK is not possible to use.
   * :kconfig:option:`CONFIG_SECURE_STORAGE_AEAD_BACKEND_KEY_CUSTOM` - Selects a custom implementation for the AEAD key provider.

Usage
*****

The trusted storage library can only be used on a build using a build target with :ref:`CMSE disabled <app_boards_spe_nspe_cpuapp>` (``_cpuapp``).
When you build for cpuapp, you build the firmware for the application core without CMSE and thus no TF-M.
The library can be used directly on such a build to store important assets.
However, for cryptographic keys we suggest to use the `PSA functions for key management`_.
These APIs will internally use this library to store persistent keys.

> 

Reviewer: We could add a note here, that external Storage is not supported by default but could be realized by implementing a custom storage backend.

Dependencies
************

This library has dependencies to following libraries:

* :ref:`lib_hw_unique_key`
* :ref:`Zephyr's settings subsystem <zephyr:settings_api>`

API documentation
*****************

Protected storage
=================

> | Header file: :file:`include/protected_storage.h`

Reviewer: We should use the headers in `subsys/encrypted_storage/include/psa/protected_storage.h` 
Instead of coping to the nrf global include folder

Author: Usually, to build the API docs, the header file is added to the include folder. Can the files be distinguished in any way, can they be named with ts_ prefix or something like that? Or is there a reason for not doing this.
There is a way of including the path to a specific file to be built, but I am checking if there are any rules/restrictions for its use.

.. trusted_storage_readme:

Trusted storage
###############

.. contents::
   :local:
   :depth: 2

The trusted storage library enables its users to provide integrity, confidentiality and authenticity of stored data using AEAD algorithms or cryptographic hash, without the use of TF-M Platform Root of Trust (PRoT).
The library implements the PSA Certified Secure Storage API.

Overview
********

Architecture
============

The trusted storage is designed and implemented in a modular way, which allows and simplifies customization.
This is realized by using backends for the different tasks performed within the trusted storage.

Trusted storage provides users two choices that are interfacing backends, ``TRUSTED_STORAGE_BACKEND`` and ``TRUSTED_STORAGE_STORAGE_BACKEND``.
The backends are also modular, allowing customization of what crypto implementation and the key nonce are provided.

> Details about the backends and their choices are described in the subsections below.

Reviewer 2: Add a link to `Choices and backends`?

The following image gives an overview of the default architecture of the trusted storage library.

.. figure:: images/trusted_storage.svg
   :alt: Overview of the trusted storage library architecture

External storage is not supported by the trusted storage library by default.

> Support for external storage can be realized by implementing a custom storage backend.

Author: Need to come back to this in improvement. Need to clarify the procedure for adding a custom backend.

   Uses a separate backend to perform the AEAD operation, set by the Kconfig option :kconfig:option:`CONFIG_TRUSTED_STORAGE_BACKEND_AEAD_CRYPTO`.
   The nonce and the key for the AEAD operation are provided by separate backends, allowing a custom implementation.

   For the key, the default choice is to use the :kconfig:option:`CONFIG_TRUSTED_STORAGE_BACKEND_AEAD_KEY_DERIVE_FROM_HUK` Kconfig option.
   With this option, a :ref:`lib_hw_unique_key` and the UID are used to derive an AEAD key.

TRUSTED_STORAGE_STORAGE_BACKEND_SETTINGS
   Stores the given assets by using :ref:`Zephyr's settings subsystem <zephyr:settings_api>`.
   The backend requires that Zephyr's settings subsystem is enabled for use (Kconfig option :kconfig:option:`CONFIG_SETTINGS` has to be set).

   The trusted storage library provides the ``TRUSTED_STORAGE_STORAGE_BACKEND_SETTINGS`` as a storage backend, but it has support for adding other memory types for storage.

Requirements
************

> Before using the trusted storage library with its default settings and options, the following must be in place:

Reviewer 2: ```suggestion
Before using the trusted storage library with its default settings and options, make sure to meet the following requirements:
```
